layer2-state-provider/history_commitment_provider.go

```diff
@@ -8,68 +8,98 @@ import (
 	"strconv"
 	"time"
 
-	protocol "github.com/offchainlabs/bold/chain-abstraction"
-	"github.com/offchainlabs/bold/state-commitments/history"
-	prefixproofs "github.com/offchainlabs/bold/state-commitments/prefix-proofs"
 	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/metrics"
 
 	"github.com/offchainlabs/bold/api"
 	"github.com/offchainlabs/bold/api/db"
 	"github.com/offchainlabs/bold/containers/option"
-	"github.com/ethereum/go-ethereum/common"
+	"github.com/offchainlabs/bold/state-commitments/history"
+
+	protocol "github.com/offchainlabs/bold/chain-abstraction"
+	prefixproofs "github.com/offchainlabs/bold/state-commitments/prefix-proofs"
 )
 
-// MachineHashCollector defines an interface which can collect hashes from an Arbitrator machine
-// at a block height, starting at a specific opcode index in the machine and stepping through it
-// in increments of custom size. Along the way, it computes each machine hash at each step
-// and outputs a list of these hashes at the end. This is a computationally expensive process
-// that is best performed if machine hashes are cached after runs.
+// MachineHashCollector defines an interface which collects hashes of the state
+// of Arbitrator machines according to a provided `cfg` argument.
+// See the documentation of the HashCollectorConfig for the details of how the
+// configuration affects the collection of machine hashes.
 type MachineHashCollector interface {
 	CollectMachineHashes(ctx context.Context, cfg *HashCollectorConfig) ([]common.Hash, error)
 }
 
-// ProofCollector defines an interface which can collect proof from an Arbitrator machine
-// at a block height, at a specific opcode index.
+// ProofCollector defines an interface which can collect a one-step proof from
+// an Arbitrator machine at a block height offset from some global state, and
+// at a specific opcode index.
 type ProofCollector interface {
 	CollectProof(
 		ctx context.Context,
-		wasmModuleRoot common.Hash,
-		fromBatch Batch,
+		assertionMetadata *AssociatedAssertionMetadata,
 		blockChallengeHeight Height,
 		machineIndex OpcodeIndex,
 	) ([]byte, error)
 }
 
-// HashCollectorConfig defines configuration options for a machine hash collector to
-// step through an Arbitrator machine at a specific L2 message number, step through it in
-// increments of `StepSize`, and collect the machine hashes at those steps into an output slice.
+// HashCollectorConfig configures the behavior the CollectMachineHashes method.
```

Comment:
Fantastic commentary. Very impressed with the information quality this file now has

Reply:
Thanks. Marking as resolved.